File: src/opentelemetry_testing.erl
-module(opentelemetry_testing).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% type definitions
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

-type attributes_map() ::
        #{opentelemetry:attribute_key() => opentelemetry:attribute_value()}.

-type event() :: #{
                   name := opentelemetry:event_name(),
                   attributes := attributes_map(),
                   term() => term()
                  }.

-type link() :: #{
                  trace_id := opentelemetry:trace_id(),
                  span_id := opentelemetry:span_id(),
                  attributes := attributes_map(),
                  term() => term()
                 }.

-type status() :: undefined |
                  #{
                    code := opentelemetry:status_code(),
                    message := unicode:unicode_binary()
                   }.

-type span_map() :: #{
                      trace_id := opentelemetry:trace_id(),
                      span_id := opentelemetry:span_id(),
                      parent_span_id := opentelemetry:span_id() | undefined,
                      name := opentelemetry:span_name(),
                      kind := opentelemetry:span_kind(),
                      status := status(),
                      attributes := attributes_map(),
                      events := [event()],
                      links := [link()],
                      start_time := opentelemetry:timestamp(),
                      end_time := opentelemetry:timestamp(),
                      term() => term()
                     }.

-type span_map_tree() :: span_collector:tree(span_map()).

-export_type([attributes_map/0,
              event/0,
              link/0,
              status/0,
              span_map/0,
              span_map_tree/0]).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% exported functions
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% API
-export([ensure_started/0,
         reset/0,
         get_span_ids_by_name/1,
         get_spans_by_name/1,
         wait_for_span/3,
         build_span_tree/2,
         convert_span/1,
         match/2]).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% API implementation
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


-spec ensure_started() -> ok.
ensure_started() ->
    application:load(opentelemetry),
    application:set_env(opentelemetry, traces_exporter, none),
    application:set_env(opentelemetry,
                        processors,
                        [{otel_simple_processor, #{}}]),
    {ok, _} = application:ensure_all_started(opentelemetry),
    ok = span_convertor:init(),
    ok = span_collector:ensure_started().


-spec reset() -> ok.
reset() ->
    span_collector:reset().


-spec get_span_ids_by_name(opentelemetry:span_name()) ->
          {ok, {opentelemetry:trace_id(), opentelemetry:span_id()}} |
          {error, not_found | span_is_not_unique}.
get_span_ids_by_name(Name) ->
    span_collector:get_span_ids_by_name(Name).


-spec get_spans_by_name(opentelemetry:span_name()) ->
          [span_map()].
get_spans_by_name(Name) ->
    Spans = span_collector:get_spans_by_name(Name),
    [ convert_span(Span) || Span <- Spans ].


-spec wait_for_span(opentelemetry:trace_id(),
                    opentelemetry:span_id(),
                    non_neg_integer()) ->
          {ok, span_map()} | {error, timeout | span_is_not_unique}.
wait_for_span(TraceId, SpanId, Timeout) ->
    case span_collector:wait_for_span(TraceId, SpanId, Timeout) of
        {ok, Span} -> {ok, convert_span(Span)};
        Ret -> Ret
    end.


-spec build_span_tree(opentelemetry:trace_id(), opentelemetry:span_id()) ->
          {ok, span_map_tree()} | {error, not_found | span_is_not_unique}.
build_span_tree(TraceId, SpanId) ->
    span_collector:build_span_tree(TraceId, SpanId, fun convert_span/1).


-spec convert_span(span_collector:span()) -> span_map().
convert_span(Span) ->
    ConversionFunctions = [fun span_convertor:records_to_maps/1,
                           fun maybe_simplify_attributes/1,
                           fun maybe_simplify_events/1,
                           fun maybe_simplify_links/1],
    FoldFn = fun(Fn, Acc) when is_function(Fn, 1) -> Fn(Acc) end,
    lists:foldl(FoldFn, Span, ConversionFunctions).


-spec match(span_matcher:value(), span_matcher:pattern()) ->
          span_matcher:match_result().
match(Value, Pattern) ->
    span_matcher:match(Value, Pattern).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% local functions
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


maybe_simplify_attributes(#{
                            attributes := #{
                                            '$record_name' := attributes,
                                            map := Attributes
                                           }
                           } = Map) ->
    %% convert attributes to a format similar to
    %% opentelemetry:attributes_map() type
    Map#{attributes := Attributes};
maybe_simplify_attributes(Term) -> Term.


maybe_simplify_events(#{
                        events := #{
                                    '$record_name' := events,
                                    list := Events
                                   }
                       } = Map) ->
    %% convert events to a format similar to opentelemetry:event() type
    SimpleEvents = [ maybe_simplify_attributes(E) || E <- Events ],
    Map#{events := SimpleEvents};
maybe_simplify_events(Term) -> Term.


maybe_simplify_links(#{
                       links := #{
                                  '$record_name' := links,
                                  list := Links
                                 }
                      } = Map) ->
    %% convert links to a format similar to opentelemetry:link() type
    SimpleLinks = [ maybe_simplify_attributes(L) || L <- Links ],
    Map#{links := SimpleLinks};
maybe_simplify_links(Term) -> Term.

File: lib/opentelemetry_testing.ex
defmodule OpentelemetryTesting do
  require Record

  @fields Record.extract(:span_ctx, from_lib: "opentelemetry_api/include/opentelemetry.hrl")
  Record.defrecordp(:span_ctx, @fields)

  #########################################################################
  ## local macros
  #########################################################################

  defmacrop raise_error(error) do
    quote do
      {function, arity} = __ENV__.function

      error_msg =
        inspect(unquote(error), limit: :infinity, printable_limit: :infinity, pretty: true)
        |> then(&Regex.replace(~r"\n", &1, "\n\t\t"))

      raise "#{__MODULE__}.#{function}/#{arity}:\n\t\t#{error_msg}"
    end
  end

  #########################################################################
  ## API implementation
  #########################################################################
  @spec ensure_started() :: :ok
  def ensure_started do
    ## when running 'mix test,' dependency applications are normally started
    ## before test execution, so the test configuration for the :opentelemetry
    ## app should be provided in the "config/config.exs" file
    :ok = :span_collector.ensure_started()
  end

  @spec reset() :: :ok
  defdelegate reset, to: :opentelemetry_testing

  @spec get_span_ids_by_name(:opentelemetry.span_name()) ::
          {:ok, {:opentelemetry.trace_id(), :opentelemetry.span_id()}}
          | {:error, :not_found | :span_is_not_unique}
  defdelegate get_span_ids_by_name(name), to: :opentelemetry_testing

  @spec get_spans_by_name(:opentelemetry.span_name()) ::
          [:opentelemetry_testing.span_map()]
  defdelegate get_spans_by_name(name), to: :opentelemetry_testing

  @spec wait_for_span(:opentelemetry.trace_id(), :opentelemetry.span_id(), non_neg_integer()) ::
          {:ok, :opentelemetry_testing.span_map()}
          | {:error, :timeout | :span_is_not_unique}
  defdelegate wait_for_span(trace_id, span_id, timeout), to: :opentelemetry_testing

  @spec build_span_tree(:opentelemetry.trace_id(), :opentelemetry.span_id()) ::
          {:ok, :opentelemetry_testing.span_map_tree()}
          | {:error, :not_found | :span_is_not_unique}
  defdelegate build_span_tree(trace_id, span_id), to: :opentelemetry_testing

  @spec match(:span_matcher.value(), :span_matcher.pattern()) :: :span_matcher.match_result()
  defdelegate match(value, pattern), to: :opentelemetry_testing

  @spec get_span_ids_by_name!(:opentelemetry.span_name()) ::
          {:opentelemetry.trace_id(), :opentelemetry.span_id()}
  def get_span_ids_by_name!(name) do
    case get_span_ids_by_name(name) do
      {:ok, data} -> data
      {:error, error} -> raise_error(error)
    end
  end

  @spec wait_for_span!(:opentelemetry.trace_id(), :opentelemetry.span_id(), non_neg_integer()) ::
          :opentelemetry_testing.span_map()
  def wait_for_span!(trace_id, span_id, timeout) do
    case wait_for_span(trace_id, span_id, timeout) do
      {:ok, data} -> data
      {:error, error} -> raise_error(error)
    end
  end

  @spec build_span_tree!(:opentelemetry.trace_id(), :opentelemetry.span_id()) ::
          :opentelemetry_testing.span_map_tree()
  def build_span_tree!(trace_id, span_id) do
    case build_span_tree(trace_id, span_id) do
      {:ok, data} -> data
      {:error, error} -> raise_error(error)
    end
  end

  @spec match!(:span_matcher.value(), :span_matcher.pattern()) :: true
  def match!(value, pattern) do
    case match(value, pattern) do
      true -> true
      {false, error} -> raise_error(error)
    end
  end

  @spec get_span_ids() :: {:opentelemetry.trace_id(), :opentelemetry.span_id()} | :undefined
  def get_span_ids do
    case OpenTelemetry.Tracer.current_span_ctx() do
      span_ctx(trace_id: trace_id, span_id: span_id) -> {trace_id, span_id}
      :undefined -> :undefined
    end
  end
end

defmodule TestModule do
  def test_function do
    :ok
  end
end
